### PR TITLE
perf: use static buffer

### PR DESCRIPTION
**Pull request by ronag**

### CODE REVIEW

**Comment by mcollina** [comment]
what's the perf benefit?

**Comment by ronag** [comment]
Less memory usage and allocations

**Comment by mcollina** [comment]
I mean, is it measurable?

Anyway, tests are failing.

**Comment by ronag** [comment]
It

> I mean, is it measurable?
> 
> Anyway, tests are failing.

Socket high watermark is 16k. So it’s 16x number of connections -1 less memory used.

In theory I guess it could be faster since the memory address stays in CPU cache. But that’s not primary point.

Also simpler code IMO.

I’ll sort the tests.

File: lib/client.js [hunk 1]
```diff
@@ -882,7 +872,15 @@ function onParserTimeout (parser) {
 
 function onSocketData (data) {
   const { [kParser]: parser } = this
-  parser.execute(data)
+
+  if (data.length <= currentBufferSize) {
+    parser.execute(data)
+  } else {
+    for (let n = 0; n < data.length; n += currentBufferSize) {
+      parser.execute(data.slice(n, currentBufferSize))
```
Comment (ronag) [inline lib/client.js:880]:
```suggestion
      parser.execute(data.slice(n, n + currentBufferSize))
```

File: lib/client.js [hunk 2]
```diff
@@ -882,7 +872,15 @@ function onParserTimeout (parser) {
 
 function onSocketData (data) {
   const { [kParser]: parser } = this
-  parser.execute(data)
+
+  if (data.length <= currentBufferSize) {
+    parser.execute(data)
+  } else {
+    for (let n = 0; n < data.length; n += currentBufferSize) {
```
Comment (ronag) [inline lib/client.js:879]:
```suggestion
    for (let n = 0; data.length; n += currentBufferSize) {
```

File: lib/client.js [hunk 3]
```diff
@@ -882,7 +872,15 @@ function onParserTimeout (parser) {
 
 function onSocketData (data) {
   const { [kParser]: parser } = this
-  parser.execute(data)
+
+  if (data.length <= currentBufferSize) {
+    parser.execute(data)
+  } else {
+    for (let n = 0; data.length; n += currentBufferSize) {
+      parser.execute(data.slice(n, n + currentBufferSize))
```
Comment (ronag) [inline lib/client.js:880]:
```suggestion
      parser.execute(data.slice(0, currentBufferSize))
```

File: lib/client.js [hunk 4]
```diff
@@ -882,7 +872,15 @@ function onParserTimeout (parser) {
 
 function onSocketData (data) {
   const { [kParser]: parser } = this
-  parser.execute(data)
+
+  if (data.length <= currentBufferSize) {
+    parser.execute(data)
+  } else {
+    for (let n = 0; data.length; n += currentBufferSize) {
```
Comment (ronag) [inline lib/client.js:879]:
```suggestion
    while (data.length) {
```

**Comment by ronag** [comment]
@mcollina PTAL

**Review by dnlup — APPROVED** [review]

**Pull request merged**